crates/jstz_api/src/file/blob.rs

```diff
@@ -0,0 +1,511 @@
+        }
+    }
+    // 2. Convert every character in relativeContentType to ASCII lowercase.
+    t.to_ascii_lowercase()
+}
+
+fn utf8_encoding(s: &[u16]) -> String {
+    char::decode_utf16(s.iter().copied())
+        .map(|r| match r {
+            Ok(c) => String::from(c),
+            Err(_) => String::from("\u{FFFD}"),
+        })
+        .collect()
+}
+
+// https://w3c.github.io/FileAPI/#process-blob-parts
+fn process_blob_parts(
+    blob_parts: BlobParts,
+    options: &Option<BlobPropertyBag>,
+    context: &mut Context<'_>,
+) -> Vec<u8> {
+    let BlobParts(blob_parts) = blob_parts;
+    // 1. Let bytes be an empty sequence of bytes.
+    let mut bytes: Vec<u8> = vec![];
+    // 2. For each element in parts:
+    for element in blob_parts.iter() {
+        match element {
+            // 1. If element is a USVString, run the following substeps:
+            BlobPart::String(string) => {
+                // 1. Let s be element.
+                let s = string.as_slice();
+                // 2. If the endings member of options is "native", set s to the result
+                //    of converting line endings to native of element.
+                let s = match options {
+                    Some(BlobPropertyBag {
+                        endings: Some(Endings::Native),
+                        type_: _,
+                    }) => convert_line_endings_to_native(s),
+                    _ => s.to_vec(),
+                };
+                // 3. Append the result of UTF-8 encoding s to bytes.
+                bytes.append(&mut utf8_encoding(s.as_slice()).into())
+            }
+            // 2. If element is a BufferSource, get a copy of the bytes held by the
+            //    buffer source, and append those bytes to bytes.
+            BlobPart::BufferSource(buffer_source) => {
+                let buffer_data = buffer_source.to_array_buffer_data(context).unwrap();
+                let mut b: Vec<u8> =
+                    buffer_data.as_slice().unwrap().iter().copied().collect();
+                bytes.append(&mut b)
+            }
+            // 3. If element is a Blob, append the bytes it represents to bytes.
+            BlobPart::Blob(blob) => bytes.append(&mut blob.bytes.clone()),
+        }
+    }
+    // 3. Return bytes.
+    bytes
+}
+
+impl Blob {
+    // https://w3c.github.io/FileAPI/#constructorBlob
+    pub fn new(
+        blob_parts: Option<BlobParts>,
+        options: &Option<BlobPropertyBag>,
+        context: &mut Context<'_>,
+    ) -> JsResult<Self> {
+        // 1. If invoked with zero parameters, return a new Blob object consisting of
+        //    0 bytes, with size set to 0, and with type set to the empty string.
+        if blob_parts.is_none() && options.is_none() {
+            return Ok(Self {
+                bytes: vec![],
+                type_: String::from(""),
+                size: 0,
+            });
+        }
+        let blob_parts = blob_parts.unwrap();
```

Comment:
Using a `match` here would be more idiomatic

Reply:
I was not sure about this one, I have updated it using a `match` for now, and giving an error in case that `options` is `Some` but `blob_parts` is `None`.

Reply:
lgtm